BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m

```diff
@@ -0,0 +1,297 @@
+{
+    NSError *error;
+
+    //persist sessions/$backgroundSessionId/$sessionTaskId
+    BOOL success = [self createDirForBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId error:&error];
+
+    if (success == YES && error == nil) {
+        NSString *path = [self dirPathOfSessionTaskWithBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+        //persist data to sessions/$backgroundSessionId/$sessionTaskId/$fileType
+
+        switch (type) {
+            case BOXURLSessionTaskCacheFileTypeDestinationFilePath:
+                path = [path stringByAppendingPathComponent:BOXURLSessionTaskCacheDestinationFilePath];
+                break;
+            case BOXURLSessionTaskCacheFileTypeResumeData:
+                path = [path stringByAppendingPathComponent:BOXURLSessionTaskCacheResumeData];
+                break;
+            case BOXURLSessionTaskCacheFileTypeResponse:
+                path = [path stringByAppendingPathComponent:BOXURLSessionTaskCacheResponse];
+                break;
+            case BOXURLSessionTaskCacheFileTypeResponseData:
+                path = [path stringByAppendingPathComponent:BOXURLSessionTaskCacheResponseData];
+                break;
+            case BOXURLSessionTaskCacheFileTypeError:
+                path = [path stringByAppendingPathComponent:BOXURLSessionTaskCacheError];
+                break;
+        }
+
+        success = [self cacheAndEncryptData:data atPath:path error:&error];
+    }
+
+    if (outError != nil) {
+        (*outError) = error;
+    }
+    return success;
+}
+
+- (BOOL)cacheAndEncryptData:(NSData *)data atPath:(NSString *)path error:(NSError **)outError
+{
+    NSData *finalData = data;
+    if ([self.delegate respondsToSelector:@selector(encryptData:)]) {
+        finalData = [self.delegate encryptData:data];
+    }
+    return [finalData writeToFile:path options:NSDataWritingWithoutOverwriting error:outError];
+}
+
+// create if not exists, users/$userId/$associateId/$backgroundSessionId-$sessionTaskId
+- (BOOL)createFileForUserId:(NSString *)userId associateId:(NSString *)associateId backgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId error:(NSError **)error
+{
+    NSString *path = [self filePathOfUserSessionTaskGivenUserId:userId associateId:associateId backgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+    return [self createFile:path error:error];
+}
+
+// create if not exists, sessions/$backgroundSessionId/$sessionTaskId
+- (BOOL)createDirForBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId error:(NSError **)error
+{
+    NSString *path = [self dirPathOfSessionTaskWithBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+    return [self createDirectory:path error:error];
+}
+
+- (BOOL)createDirectory:(NSString *)path error:(NSError **)error
+{
+    BOOL isDir = NO;
+    BOOL success = YES;
```

Comment:
If it already exists, it should return NO IMO. Technically, it did *not* create the directory this time, which is what the method says.